src/main/java/io/vertx/core/impl/KeySeq.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.core.impl;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+class KeySeq {
```

Comment:
I think that this one should be an assigned field to a context, because not all contexes requires the same "key space", or we risk to have every context to always allocate N keys even if they requires very different number of keys.

The other option, instead is to make things smarter working in chunks (and lazy):
- no `KeySeq`, but a chunked approach with N (eg 8) elements per chunk saved in a CopyOnWriteArrayList<KeyChunk> keys;
- a lookup in position N will hash N to find the proper bucket in the key space int chunkIndex = N / CHUNKS_CAPACITY and keyOffset = N % CHUNKS_CAPACITY
- each chunk is allocated lazily based on which index has been accessed